Add SSO RFC

stevenvegt opened the pull request:
How to do an SSO with Nuts.
This adds an RFC for the SSO Bolt.

Todo:
* [x] Update Getting started with Nuts - Setup local nodes guide
* [x] Remove getting started section from RFC

stevenvegt committed:
Add skeleton version of the SSO rfc

reinkrul commented:
I can hardly argue against this change (since there's not much content yet), other than this filename uses dashes (-rfc) and the previous one (session tokens) uses underscores (_rfc)...

stevenvegt committed:
Add basic steps and entpoint type

stevenvegt committed:
Fix list numbering

stevenvegt committed:
Fix layout

stevenvegt committed:
Add more TODO items

stevenvegt commented:
@ChrisBrandhorst Will this work for you?

ChrisBrandhorst commented:
@stevenvegt In essence, yes, but a bit shallow (but WIP of course). Questions:

- Is "Authorization Server" a Nuts or a local concept?
- Jump URL _must_ be on the Auth Server?
- We're not clear on the "care provider jump URL (registration)" part. Please elaborate on that.

Also, a number of times requests are mentioned, but no details are specified which endpoints are used, what parameters are required, etc.

stevenvegt commented:
* The _Authorization Server_ is a local concept and has to be implemented by the destination vendor. The Authorization Server needs information about the inner workings of the destination application, like how does the URL for the patients dossier looks like. So there will probably be some mapping of BSN -> PatientID.
Also, the Authorization Server needs to create an internal session for the user on the destination application. It even might need to create an account and perform an email -> accountID mapping.

* Yes, The users _Session Token_ should some how arrive at this Authorization Server.

* We use a Nuts registry (part of the Nuts node) to announce endpoints. This is a simple json hash with a type and location. The registration is currently just adding this endpoint to the github repository:  https://github.com/nuts-foundation/nuts-registry-development/blob/master/endpoints.json

stevenvegt committed:
Provide more details to each step

References to Api endpoints and such.

stevenvegt committed:
Remove consent store checks

reinkrul commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,106 @@
+Jump
+====
+
+The user gets a status 303 or 303 with the jump sso url including the session token.
```
I think you mean 302 or 303?

reinkrul commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,106 @@
+
+The application the user jumps to can only show patient provided within the session.
+
+The application the user jumps to does not get notified about logout operations
```
in other words, there is no Global Logout feature (local only)

stevenvegt commented:
Add a note about encrypting the jump token, since there could be a BSN inside.

stevenvegt committed:
Add API calls in cUrl format

stevenvegt committed:
Fix http status code

stevenvegt committed:
Add token encryption todo

stevenvegt committed:
Add fetch endpoint requests

stevenvegt committed:
Several improvements

* Add terminology section
* Use consistant wording like access token intead of session tokens, client instead of source application
* Add Get endpoint Api calls

reinkrul commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,230 @@
+The an registry endpoint should use the following for the Nuts OAuth authorization server
+::
+
+  urn:ietf:rfc:3986:urn:oid:1.3.6.1.4.1.54851.1:nuts-oauth-authentication-server
```
authentication or authorization server?

woutslakhorst commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,230 @@
+
+   .. code-block:: console
+
+     $ curl --location --request GET 'localhost:1323/api/endpoints?orgIds=urn:oid:2.16.840.1.113883.2.4.6.1:00000001&type=urn:ietf:rfc:3986:urn:oid:1.3.6.1.4.1.54851.1:nuts-sso-jump-endpoint'
```
Double urn? Is that needed by the spec?

stevenvegt replied:
I'm not sure. The `urn:ietf:rfc:3986` is the system and the `oid:1.3.6.1.4.1.54851.1:nuts-sso-jump-endpoint` the value. So, probably without the second `urn` like this: `urn:ietf:rfc:3986:oid:1.3.6.1.4.1.54851.1:nuts-sso-jump-endpoint`

woutslakhorst replied:
I meant type is defined by our spec as it must be a URI, the identifier is already a URI:
```
urn:oid:1.3.6.1.4.1.54851.1:nuts-sso-jump-endpoint
```
right now it is something like: URI is of type URI is of type URI is of type URI....

woutslakhorst commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,230 @@
+
+   .. code-block:: console
+
+     $ curl --location --request POST 'localhost:1323/auth/contract/session' \
```
Comment about future addition for adding scope?

stevenvegt replied:
👍 I will add a `scope` field to the `CreateJwtBearerToken` request.

woutslakhorst commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,230 @@
+
+Determine which IRMA attributes are needed.
+
+Do we need a separate Jump contract?
```
Could be scope, as mentioned before

woutslakhorst commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,230 @@
+
+  urn:ietf:rfc:3986:urn:oid:1.3.6.1.4.1.54851.1:nuts-sso-jump-endpoint
+
+TODO
```
Maybe combine 2 API's into one? (login contract + jump token)

stevenvegt replied:
The login contract consists of createContractSession and getContractStatus. This status might return the signed irma contract, but might not if the signing failed. Due to this splitted nature, I do not see an elegant way of combining the login contract with obtaining the jump token.

stevenvegt committed:
Fix wrong name of authorization server

stevenvegt committed:
Fix urns

ChrisBrandhorst commented:
- Step 11.i says "Retrieve the **access token** from the **access token**", step 10 mentions "session_token", please clarify
- Steps 11.i and 11.ii are still somewhat obscure. Which token contains the identity of the user? Why must I use the Nuts Auth to validate a token I received from that same node?

stevenvegt committed:
Several updates

* Rename session tokens to access tokens
* Add curl request for token introspection
* Fix some code identation

stevenvegt commented:
> * Step 11.i says "Retrieve the **access token** from the **access token**", step 10 mentions "session_token", please clarify

Yes, this was a bit confusing. We first introduced the concept of sessions which we later rewrote to just access tokens in relation to a request context. I fixed this.

> * Steps 11.i and 11.ii are still somewhat obscure. Which token contains the identity of the user? Why must I use the Nuts Auth to validate a token I received from that same node?

The access token is opaque to both the client and resource server. For the resource server, the request context can be retrieved by posting the token to the authorization server which does something called "token introspection" returning a JSON document with all the relevant data, including the "active" field.
I rewrote this part and added an introspection cUrl call. Let me know if this is still unclear.

JoeriLandman commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,245 @@
+
+#. The client application renders a SSO button
+#. The User clicks the button
+#. The client application collects the users identity using a login contract. If not already present, it lets the user sign one using IRMA
```
```suggestion
#. The client application collects the users identity using a login contract. If not already present, it lets the user sign one using IRMA.
```

JoeriLandman commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,245 @@
+        }]
+
+#. The client application renders a SSO button
+#. The User clicks the button
```
```suggestion
#. The User clicks the button.
```

JoeriLandman commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,245 @@
+          "version": "0.1.0"
+        }]
+
+#. The client application renders a SSO button
```
```suggestion
#. The client application renders a SSO button.
```

JoeriLandman reviewed (commented):
Ending the sentences :)

JoeriLandman commented on pages/rfc/sso-rfc.rst:
```diff
@@ -0,0 +1,245 @@
+#. The User loads a page in the client application in the context of a certain subject (patient).
+#. The client application checks rights and settings to see if this user is allows to jump (local policy)
+#. The client application checks if the subject has any external custodians (care providers). This subject <-> custodian relationship should be registered locally.
+#. For each custodian the client application checks if it has a jump endpoint
```
```suggestion
#. For each custodian the client application checks if it has a jump endpoint.
```

JoeriLandman commented:
Steps 1 till 6 in `SSO Steps` are steps required for UX (showing the button). This is a good example of a flow through a client application. However you might want to change the `SSO Steps` to something like `SSO Steps Example`. Since these steps are not all required to do the SSO jump.

stevenvegt committed:
Add new tokens with IRMA attributes

Rename endpoint type

stevenvegt committed:
Fix url in example call

stevenvegt committed:
Update the rfc documentation

Updates the new registry endpoint type with authorizationURL embedded.
Adds steps to setup a local network with demo-ehr and how to register a consent and make a jump.

stevenvegt committed:
Fix port number

stevenvegt committed:
Fix nuts-sso endpoint

stevenvegt committed:
Add better example for the endpoint

stevenvegt committed:
Clean up some text

stevenvegt committed:
Update endpoints section in accessToken rfc

Authorization server is no longer its own endpoint but provided along the resource server registry endpoint.

stevenvegt committed:
Fix command

stevenvegt committed:
Clear up example calls

use self-explaining domain names instead of confusing localhost names with ports that change if you run on docker of using the binary.

reinkrul committed:
Linux compatibility

stevenvegt committed:
Update getting started guide

stevenvegt committed:
Merge branch 'master' into add-sso-rfc

stevenvegt committed:
Finishing touches

stevenvegt committed:
remove duplicate instructions from RFC
